File: _recipes/python/healthcheck.py
# Recipe: /health and /ready HTTP handlers, stdlib only
# Stdlib-only.
# Usage:
#   from http.server import HTTPServer
#   from healthcheck import HealthHandler, set_ready
#
#   server = HTTPServer(("0.0.0.0", 8080), HealthHandler)
#   set_ready(True)
#   server.serve_forever()
#
# /health  -> 200 {"status":"ok"} always
# /ready   -> 200 {"status":"ready"} or 503 {"status":"not_ready"} based on set_ready()
from __future__ import annotations

import json
import threading
from http.server import BaseHTTPRequestHandler

_ready_lock = threading.Lock()
_ready = False


def set_ready(value: bool) -> None:
    """Mark the service as ready (or not). Thread-safe."""
    global _ready
    with _ready_lock:
        _ready = bool(value)


def _is_ready() -> bool:
    with _ready_lock:
        return _ready


class HealthHandler(BaseHTTPRequestHandler):
    def do_GET(self) -> None:
        if self.path == "/health":
            self._json(200, {"status": "ok"})
        elif self.path == "/ready":
            if _is_ready():
                self._json(200, {"status": "ready"})
            else:
                self._json(503, {"status": "not_ready"})
        else:
            self._json(404, {"status": "not_found", "path": self.path})

    def _json(self, status: int, payload: dict[str, object]) -> None:
        body = json.dumps(payload).encode()
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def log_message(self, format: str, *args: object) -> None:  # noqa: A002
        return

File: _recipes/node/healthcheck.ts
// Recipe: /health and /ready HTTP handlers, stdlib only
// Stdlib-only (node:http).
// Usage:
//   import { createHealthServer, setReady } from './healthcheck.js';
//   const server = createHealthServer();
//   server.listen(8080);
//   setReady(true);
//
// /health  -> 200 {"status":"ok"} always
// /ready   -> 200 {"status":"ready"} or 503 {"status":"not_ready"}

import { createServer, type Server } from 'node:http';

let ready = false;

export function setReady(value: boolean): void {
  ready = Boolean(value);
}

export function createHealthServer(): Server {
  return createServer((req, res) => {
    const send = (status: number, payload: Record<string, unknown>): void => {
      const body = JSON.stringify(payload);
      res.writeHead(status, {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(body),
      });
      res.end(body);
    };

    if (req.url === '/health') {
      send(200, { status: 'ok' });
    } else if (req.url === '/ready') {
      if (ready) send(200, { status: 'ready' });
      else send(503, { status: 'not_ready' });
    } else {
      send(404, { status: 'not_found', path: req.url });
    }
  });
}
